Fixed #32002 -- Added headers parameter to HttpResponse and subclasses.

knyghty opened the pull request:
[Ticket](https://code.djangoproject.com/ticket/32002)

There are now two overlapping ways to set the `Content-Type` header. I think it makes sense to take `content_type` over `headers`, but not sure if we should check this in `__init__()` and raise an exception, document it, or perhaps neither is necessary.

charettes commented on docs/howto/outputting-csv.txt:
```diff
@@ -20,8 +20,10 @@ Here's an example::
-        response.headers['Content-Disposition'] = 'attachment; filename="somefilename.csv"'
+        response = HttpResponse(
+            content_type='text/csv',
+            headers={'Content-Dosposition': 'attachment; filename="somefilename.csv"'},
```
```suggestion
            headers={'Content-Disposition': 'attachment; filename="somefilename.csv"'},
```

matthiask commented on docs/topics/class-based-views/index.txt:
```diff
@@ -117,9 +117,10 @@ And the view::
         def head(self, *args, **kwargs):
             last_book = self.get_queryset().latest('publication_date')
-            response = HttpResponse()
             # RFC 1123 date format
```
This comment should be moved above the `headers=` argument.

knyghty replied:
I think neither place really feels right, but I think it's slightly better there.

matthiask commented on tests/httpwrappers/tests.py:
```diff
@@ -470,6 +470,10 @@ def test_header_deletion(self):
 
+    def test_instantiate_with_headers(self):
+        r = HttpResponse('hello', headers={'X-Foo': 'foo'})
+        self.assertEqual(r.headers['X-Foo'], 'foo')
```
Maybe test whether `r.headers['x-foo']` is set or something? (`CaseInsensitiveMapping` itself is probably thoroughly tested but it may be useful to verify the case insensitivity here too)

knyghty replied:
I'm not exactly sure what it's protecting against, but seems reasonable, have done this.

matthiask commented:
Looks good! 

I don't want to complicate things more in this pull request BUT have you considered adding support for this argument to `TemplateResponse` and `TemplateSimpleResponse` too? Those two classes use explicit forwarding of arguments, not `*a, **kw`. I think Carlton definitely has a point regarding not adding a headers parameter to the `render()` shortcut, but the template responses are a different case.

knyghty commented:
I'm not against it but I'll hold off for a second opinion.

Is there some reason those classes don't just pass `*args, **kwargs` through like the other subclasses do?

matthiask commented:
Not passing `*args` makes sense since it's allowed to not specify the context at all.

Regarding not passing `**kwargs`: I do not see a good reason for this.

felixxm commented:
> There are now two overlapping ways to set the Content-Type header. I think it makes sense to take content_type over headers, but not sure if we should check this in __init__() and raise an exception, document it, or perhaps neither is necessary.

I think we should raise a `ValueError` that they're mutually exclusive, and maybe deprecate `content_type`.

felixxm reviewed (commented):
@knyghty Thanks :+1:

I think we should also update ["Telling the browser to treat the response as a file attachment"](https://docs.djangoproject.com/en/3.1/ref/request-response/#telling-the-browser-to-treat-the-response-as-a-file-attachment) section and pass `Content-Type` and `Content-Disposition` in `headers`.

carltongibson commented:
I think we should leave `content_type`. Just as before, it's a handy shortcut for (probably) the most common header. Plus **much code** will be using it. 

>I think we should raise a ValueError that they're mutually exclusive...

Yes, 👍. If `content_type is not None` we should `raise` if `Content-Type` (case-insensitive) in `headers`.

knyghty commented:
I'd love to deprecate it, but I agree with your Carlton, I think it would cause too many unnecessary complaints for too little gain.

I will add the check though, and update those docs.

@felixxm, @carltongibson, what do you both think about adding `**kwargs` to `TemplateResponse` and `SimpleTemplateResponse`? It seems like a good idea to me.

carltongibson commented:
> @felixxm, @carltongibson, what do you both think about adding **kwargs to TemplateResponse and SimpleTemplateResponse? It seems like a good idea to me.

I see the case for adding `headers` here. I'd think changing to `**kwargs` would be out of scope for this PR (and something we'd want to think about... — it's not always the case that the broader API is better...)

knyghty commented:
Thanks both, addressed those and fixed an issue where content type wouldn't be set from the headers dict. Also moved the code closer together to make it more understandable.

felixxm commented on tests/template_tests/test_response.py:
```diff
@@ -216,6 +216,15 @@ def test_pickling_cookie(self):
+            {'value': 123, 'fn': datetime.now},
+            headers={'X-Foo': 'foo'},
+        )
+        response.render()
```
Do we need to call `render()`? It looks unnecessary.

felixxm commented on tests/template_tests/test_response.py:
```diff
@@ -319,6 +328,16 @@ def test_repickling(self):
+            {'value': 123, 'fn': datetime.now},
+            headers={'X-Foo': 'foo'},
+        )
+        response.render()
```
Do we need to call `render()`? It looks unnecessary.

felixxm commented on tests/template_tests/test_response.py:
```diff
@@ -319,6 +328,16 @@ def test_repickling(self):
+            headers={'X-Foo': 'foo'},
+        )
+        response.render()
+        assert 'X-Foo' in response.headers
```
```suggestion
        self.assertEqual(response.headers['x-foo'], 'foo')
```

felixxm commented on tests/template_tests/test_response.py:
```diff
@@ -216,6 +216,15 @@ def test_pickling_cookie(self):
+            headers={'X-Foo': 'foo'},
+        )
+        response.render()
+        assert 'X-Foo' in response.headers
```
```suggestion
        self.assertEqual(response.headers['x-foo'], 'foo')
```

felixxm commented on django/template/response.py:
```diff
@@ -11,7 +11,7 @@ class SimpleTemplateResponse(HttpResponse):
 
     def __init__(self, template, context=None, content_type=None, status=None,
-                 charset=None, using=None):
+                 charset=None, using=None, headers=None):
```
[Docs](https://docs.djangoproject.com/en/3.1/ref/template-response/#django.template.response.SimpleTemplateResponse.__init__) changes are missing.

felixxm commented on django/template/response.py:
```diff
@@ -139,6 +139,6 @@ class TemplateResponse(SimpleTemplateResponse):
     def __init__(self, request, template, context=None, content_type=None,
-                 status=None, charset=None, using=None):
-        super().__init__(template, context, content_type, status, charset, using)
+                 status=None, charset=None, using=None, headers=None):
```
[Docs](https://docs.djangoproject.com/en/3.1/ref/template-response/#django.template.response.TemplateResponse.__init__) changes are missing.

felixxm commented on docs/releases/3.2.txt:
```diff
@@ -314,6 +314,9 @@ Requests and Responses
   Both interfaces will continue to be supported. See
   :ref:`setting-header-fields` for details.
 
+* The new ``headers`` parameter of the ``:class:`~django.http.HttpResponse`` class
```
```
* The new ``headers`` parameter of ``:class:`~django.http.HttpResponse``,
  :class:`~django.template.response.SimpleTemplateResponse`, and
  :class:`~django.template.response.TemplateResponse` allows ...
```

felixxm commented on tests/httpwrappers/tests.py:
```diff
@@ -470,6 +470,26 @@ def test_header_deletion(self):
+        assert r.headers['Content-Type'] == 'application/json'
+
+    def test_content_type_set_and_in_headers(self):
+        with self.assertRaisesMessage(
+            ValueError,
+            '`content_type` must not be set if `Content-Type` is set in the response headers.',
+        ):
```
```suggestion
        msg = ...
        with self.assertRaisesMessage(ValueError, msg):
```

felixxm reviewed (commented):
@knyghty Thanks for updates :+1:

knyghty commented:
Thanks @felixxm, I've resolved those and also added a test for the case I found yesterday.

felixxm commented on tests/template_tests/test_response.py:
```diff
@@ -222,8 +222,7 @@ def test_headers(self):
         )
-        response.render()
-        assert 'X-Foo' in response.headers
+        assert response.headers['X-Foo'] == 'foo'
```
Please use `assertEqual()`  instead of `assert ... == ...`.

knyghty replied:
Gah, sorry, force of habit, I'm too used to pytest these days.

felixxm commented on tests/httpwrappers/tests.py:
```diff
@@ -470,6 +470,28 @@ def test_header_deletion(self):
+
+    def test_content_type_set(self):
+        r = HttpResponse('hello', content_type='application/json')
+        assert r.headers['Content-Type'] == 'application/json'
```
This test works without the patch, I will move it to a separate commit.

felixxm reviewed (commented):
@knyghty Thanks for updates :+1: 

I added missing `versionchanged` annotations, added the ``headers`` parameter to ``HttpResponse.__init__()`` signature in docs, moved an extra test to a separate commit, and 
pushed minor edits.

carltongibson reviewed (approved):
OK, looks good. Thanks @knyghty  this'll be nice.

felixxm commented:
Merged in 2e7cc95499f758a1c4aa036cbf1dcddf82a89ea2 and dcb69043d0de45bb55998fc418d93c28bc7689ae.
